File: apps/api/src/first/first.service.ts
import { Inject, Injectable } from '@nestjs/common';
import ky, { type KyInstance } from 'ky';
import { ConfigService } from '../config/config.service';
import { MatchLevel } from './enums/match-level.enum';
import type { FrcEventMatchScores } from './interfaces/frc-event-scores.interface';
import type { FrcSchedule } from './interfaces/frc-schedule.interface';

@Injectable()
export class FirstService {
	private readonly http: KyInstance;

	constructor(@Inject(ConfigService) configService: ConfigService) {
		const basicAuthToken = Buffer.from(
			`${configService.frcEventsApi.username}:${configService.frcEventsApi.password}`,
			'utf8',
		).toString('base64');

		this.http = ky.extend({
			headers: {
				authorization: `Basic ${basicAuthToken}`,
			},
			prefixUrl: 'https://frc-api.firstinspires.org/v3.0',
		});
	}

	async listEventScores(year: number, eventCode: string, level: MatchLevel): Promise<FrcEventMatchScores> {
		const response = await this.http.get(
			`${encodeURIComponent(year)}/scores/${encodeURIComponent(eventCode)}/${encodeURIComponent(level)}`,
		);

		return response.json<FrcEventMatchScores>();
	}

	async getSchedule(year: number, eventCode: string): Promise<FrcSchedule> {
		const qualSearchParams = new URLSearchParams({ tournamentLevel: MatchLevel.Qualification });
		const playoffSearchParams = new URLSearchParams({ tournamentLevel: MatchLevel.Playoff });

		const [qualResponse, playoffResponse] = await Promise.all([
			this.http.get(`${encodeURIComponent(year)}/schedule/${encodeURIComponent(eventCode)}?${qualSearchParams}`),
			this.http.get(`${encodeURIComponent(year)}/schedule/${encodeURIComponent(eventCode)}?${playoffSearchParams}`),
		]);

		const [qual, playoff] = await Promise.all([qualResponse.json<FrcSchedule>(), playoffResponse.json<FrcSchedule>()]);

		return {
			// biome-ignore lint/style/useNamingConvention: This is an external API
			Schedule: [...qual.Schedule, ...playoff.Schedule],
		};
	}
}

File: apps/api/app/first/first_service.py
import asyncio
import httpx
from app.config import frc_events_api_key, frc_events_username

from .types import FrcEventMatchScores, FrcSchedule, FrcMatchLevel


class FirstService:
    """Handles communication with the FRC Events API (FIRST Inspires)."""

    BASE_URL = "https://frc-api.firstinspires.org/v3.0"

    def __init__(
        self, username: str = frc_events_username, api_key: str = frc_events_api_key
    ):
        self.auth = httpx.BasicAuth(username, api_key)

    async def list_event_scores(
        self, year: int, event_code: str, level: FrcMatchLevel
    ) -> FrcEventMatchScores:
        async with httpx.AsyncClient(
            base_url=self.BASE_URL, auth=self.auth
        ) as first_client:
            response = await first_client.get(
                f"/{year}/scores/{event_code}/{level.value}"
            )
            response.raise_for_status()

            return FrcEventMatchScores.model_validate_json(response.text)

    async def get_schedule(self, year: int, event_code: str) -> FrcSchedule:
        async with httpx.AsyncClient(
            base_url=self.BASE_URL, auth=self.auth
        ) as first_client:
            qual_response, playoff_response = await asyncio.gather(
                first_client.get(
                    f"/{year}/schedule/{event_code}",
                    params={"tournamentLevel": FrcMatchLevel.QUALIFICATION.value},
                ),
                first_client.get(
                    f"/{year}/schedule/{event_code}",
                    params={"tournamentLevel": FrcMatchLevel.PLAYOFF.value},
                ),
            )

            qual_response.raise_for_status()
            playoff_response.raise_for_status()

            qual_schedule = FrcSchedule.model_validate_json(qual_response.text)
            playoff_schedule = FrcSchedule.model_validate_json(playoff_response.text)

            return FrcSchedule(
                Schedule=qual_schedule.schedule + playoff_schedule.schedule
            )
